# Importing libraries

In [171]:
#Import libraries
import pandas as pd
import numpy as np
import os

# Importing Data

In [172]:
# Turning project folder path into a string
path=r'/Users/jasmeetakaur/Downloads/Jasmeeta/Data Immersion Course/Achievement 1/Achievement 4/Instacart Basket Analysis'

In [173]:
# Importing the “orders.csv” file
df_ords=pd.read_csv(os.path.join(path,'02 Data/Prepared Data/orders_wrangled.csv'), index_col=False)

In [174]:
# Importing the products.csv file
df_prods=pd.read_csv(os.path.join(path,'02 Data/Original Data/products.csv'), index_col=False)

# Steps for consistency check

## Mixed data

In [175]:
# Create a dataframe
df_test=pd.DataFrame()

In [176]:
# Create a mix type column
df_test['mix']=['a','b',1,True]

In [177]:
# Listing the columns of df_test
df_test.head()

mix
0     a
1     b
2     1
3  True

In [178]:
# Checking for mixed data types
for col in df_test.columns.tolist():
     weird=(df_test[[col]].applymap(type) !=df_test[[col]].iloc[0].apply(type)).any(axis=1) 
     if len (df_test[weird]) > 0:
            print (col)

mix


In [179]:
# Changing 'mix' to string type:
df_test['mix'] = df_test['mix'].astype('str')

In [180]:
# Rechecking for mixed data types
for col in df_test.columns.tolist():
     weird=(df_test[[col]].applymap(type) !=df_test[[col]].iloc[0].apply(type)).any(axis=1) 
     if len (df_test[weird]) > 0:
            print (col)
     else: print ('same type')   

same type


## Taking care of missing values

In [181]:
# To find the sum of missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [182]:
# Creating a subset with the missing values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [183]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [184]:
# Checking the number of rows and columns in df_prods
df_prods.shape

(49693, 5)

In [185]:
# Creating a new data frame excluding the missing values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False] 

In [186]:
# Identifying the number of rows and columns in df_prods_clean
df_prods_clean.shape

(49677, 5)

## Taking care of duplicate values

In [187]:
# creating a new subset containing only the duplicate rows
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [188]:
# creating a new dataframe that doesn’t include the duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [189]:
# Identifying the number of rows and columns in df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5 Data Consistency Checks

### Step 2: Generating descriptive statistics for the numeric values in df_ords dataframe


In [190]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

#### As far as df_ords is concerned nothing obviously pops out. min, max and mean appear within the reasonable range

## Data consistency check for df_prods

In [191]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

### In df_prods, the 'max price' of 99,999 looks a little off.
### Even more so when the mean value is 9.99. 
### This could be due to a problem with data entry, like misplacement of a period.
### It can be resolved by correcting the placement of the period.

### Replacing the max value of 99999 with the mean value of 9.99

In [192]:
df_prods['prices'].replace(99999, 9.99, inplace=True)
df_prods['prices'].replace(14900, 14.90, inplace=True)

In [193]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682460
std    14343.717401     38.316774       5.850282      4.200295
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

### Step 3: Checking for mixed data type in df_ords


In [194]:
for col in df_ords.columns.tolist():
     weird=(df_ords[[col]].applymap(type) !=df_ords[[col]].iloc[0].apply(type)).any(axis=1) 
     if len (df_ords[weird]) > 0:
            print (col)
     else: print(col, 'same type')   

Unnamed: 0 same type
order_id same type
user_id same type
order_number same type
order_day_of_week same type
order_hour_of_day same type
days_since_prior_order same type


### Step 4: There is no mix data type in any of the columns. There is nothing to fix.

### Step 5: Checking for missing values in df_ords


In [195]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### There are 206,209 missing values in the column 'days_since_prior_order'.

In [196]:
# Creating a subset with the missing values from the df_ords
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

### After observing the table it appears that the missing value for the column 'days_since_prior_order' is mostly for order_number 1. It indicates that this is the first time the user has ordered.

In [197]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

### Step 6: Addressing the missing values: It does not make sense to drop these rows or impute the value with mean or median. I think the missing values should be replaced by 0 since the observation indicates that the user is ordering for the first time.

In [198]:
df_ords['days_since_prior_order']=df_ords['days_since_prior_order'].fillna(0)

In [199]:
# Verifying that NaN values do not exist anymore
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]
df_ords_nan

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### Step 7: Checking for duplicate values

In [200]:
df_ords_dups=df_ords[df_ords.duplicated()]

### Step 8: There are no duplicates found in df_ords.

In [203]:
df_ords=df_ords.drop('Unnamed')

KeyError: "['Unnamed'] not found in axis"

## Exporting cleaned df_prods and df_ords data as “.csv” files in “Prepared Data” folder

In [ ]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_consistent.csv'))

In [ ]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_consistent.csv'))